apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
+    Logger.info("Animation trigger for skill #{skill.name} ready. Creating #{Enum.count(skill.effects)} effects.")
+    current_state = trigger_skill_effects(skill, current_state, initial_step_state)
+    {%{skill | effects_trigger: -1}, current_state}
+  end
 
-  # Is cooldown ready?
-  defp can_cast_basic_skill(unit), do: unit.basic_skill.remaining_cooldown <= 0
+  # Calculate the new state of the battle after a step passes for a skill being cast, specifically for its `effects_trigger` value.
+  # If the effects have already triggered, do nothing.
+  defp process_skill_effects_trigger_value(%{effects_trigger: -1} = skill, current_state, _initial_step_state),
+    do: {skill, current_state}
 
-  defp cast_skill(unit, skill, initial_step_state, current_state) do
-    target_ids = choose_targets(unit, skill, initial_step_state)
+  # Calculate the new state of the battle after a step passes for a skill being cast, specifically for its `effects_trigger` value.
+  # If the effect hasn't triggered yet, reduce the remaining effects_trigger counter.
+  defp process_skill_effects_trigger_value(skill, current_state, _initial_step_state) do
+    {%{skill | effects_trigger: skill.effects_trigger - 1}, current_state}
+  end
 
-    Enum.reduce(skill.effects, current_state, fn effect, new_state ->
-      # If skill doesn't have targeting strategy, we fall back to the effects'
-      target_ids =
-        if is_nil(target_ids),
-          do: choose_targets(unit, effect, initial_step_state),
-          else: target_ids
+  # Calculate the new state of the battle after a step passes for a pending effect.
+  # If the effect is ready to be processed, we apply it.
+  defp process_step_for_effect(%{delay: 0} = effect, current_state) do
+    Logger.info("#{format_unit_name(effect.caster)}'s effect is ready to be processed")
 
-      targets_after_effect = Enum.map(target_ids, fn id -> apply_effect(effect, unit, new_state[id]) end)
+    targets_after_effect =
+      Enum.map(effect.targets, fn id ->
+        maybe_apply_effect(effect, current_state.units[id], effect.caster, current_state.step_number)
+      end)
 
-      Enum.reduce(targets_after_effect, new_state, fn target_unit, acc_state ->
-        Map.put(acc_state, target_unit.id, target_unit)
+    current_state =
+      Enum.reduce(targets_after_effect, current_state, fn target, acc_state ->
+        put_in(acc_state, [:units, target.id], target)
       end)
-    end)
+
+    Map.put(current_state, :pending_effects, List.delete(current_state.pending_effects, effect))
+  end
+
+  # Calculate the new state of the battle after a step passes for a pending effect.
+  # If the effect isn't ready to be processed, we reduce its remaining delay.
+  defp process_step_for_effect(effect, current_state) do
+    Map.put(current_state, :pending_effects, [
+      %{effect | delay: effect.delay - 1} | List.delete(current_state.pending_effects, effect)
+    ])
```

Comment:
I'm not sure if i'm a huge fan of this recurrent tactic of updating something from a list deleting it  and adding it later on, maybe we could reduce the effects from the game state and updating/deleting/applying them in the same reduce
```
{updated_effects, updated_game_state}
Enum.reduce(new_state.pending_effects, {[], new_state], fn effect, {result_effect, current_state} ->
  cond do 
     apply_effect -> {updated_effect, updated_game_state}
     # Rest of cases
  end
end
```

Reply:
Not a fan of it either! Your solutions is pretty ingenious, I'll work on implementing it!

Reply:
I'm quite pleased with the result, let me know what you think! https://github.com/lambdaclass/mirra_backend/pull/373/commits/6ad8ac4096727feee66b13ec09b245ec86ebb41b

Reply:
I really like the change! it's a lot easier to follow now. 
just a small comment, you can iterate over an enumerable and create a new map with `Map.new` so if you iterate the units changing them you can then merge that map with the one in the state, something like:

```
targets_after_effect =
              Map.new(effect.targets, fn id ->
               {id,
                maybe_apply_effect(effect, current_state.units[id], effect.caster, current_state.step_number)}
              end)

new_state =
              update_in(current_state, [:units], fn units -> Map.merge(units, targets_after_effect) end)
```

Reply:
Yeah that's cleaner!